src/qibo/tensorflow/gates.py

```diff
@@ -508,11 +492,11 @@ def _calculate_unitaries(self):
         return matrices, additional_matrix
 
     def _prepare(self):
-        self.cgates.VariationalLayer._prepare(self)
+        cgates.VariationalLayer._prepare(self)
```

Comment:
That's a good idea, avoids code duplication. The only downside (which we don't have so far) is if the attributes used by the external method do are not allocated/named from the original class, i.e. in this case you suppose `self.is_dagger` exists, because it is obviously allocated in the base class.

Reply:
Thanks for the review and for merging. I was actually going to reply to this, I just did not have time before. Yes, this is done to avoid code duplication, however I generally don't like it as a practice to call other class methods without strictly respecting inheritance. 

Generally, I am not really sure if the way we implement gates for different backends is the best one and what problems could arise if we attempt to add a new backend in the future. This is because the initial design did not consider that we would have the custom backend (this was added as Qibo evolved) and admittedly the focus was more on benchmarks/performance rather than code simplicity and having a well-defined inheritance structure. I was a bit loose on the latter part for the sake of getting things to work and with acceptable performance. At some point, depending on future developments, we may want to have a second look on the backends and gate inheritance and have a more well-defined structure of what is defined in base and what in each backend.

The main reason I am not spending much time on this now is that I am quite unsure about the future of Tensorflow backends. Since we added density matrix support on the custom backend the only two things that could justify using a native backend now are 1. compilation and 2. automatic differentiation, which I am not sure how useful are in applications (point 2 could be useful, however it is unclear if this approach is really useful for optimization of variational circuits). With this in mind I am not sure for how long it would be worth to maintain + complicate the code & docs for keeping these backends. On the other hand, we cannot really drop them as Qibo was officially released this way.

Reply:
Thanks for the reply, and yes, that's a good point. Indeed, we could simplify the backends and limit the implementation to custom however we still in a warm-up phase. I would like to have only the custom backend, however the sgd approach will not work out of the box. We may investigate more about https://github.com/tensorflow/tensorflow/issues/40006 or eventually replace tf with [jax](https://github.com/google/jax).